pkg/workload/schemachange/query_util.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+package schemachange
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/cockroachdb/errors"
+	"github.com/jackc/pgx/v5"
+)
+
+const (
+	descriptorsJSON = `SELECT id, crdb_internal.pb_to_json('desc', descriptor) AS descriptor FROM system.descriptor`
+)
+
+type CTE struct {
+	As    string
+	Query string
+}
+
+// With is a helper for building queries utilizing Common Table Expressions
+// (CTEs). Managing reusable and composable SQL queries in Golang is quite
+// difficult without building out a full expression engine. The denormalized
+// nature of the descriptor table, however, necessitates some degree of
+// reusability. Our solution is to provide many CTE expression as constants and
+// allow them to be composed with With.
+// Usage:
+//
+//	With([]CTE{
+//		{"descriptors", jsonifiedDescriptors},
+//		{"table_descs", jsonifiedTableDescriptors},
+//	}, `SELECT * FROM descriptors WHERE EXISTS(table_descs)`)
+func With(ctes []CTE, query string) string {
+	var b strings.Builder
+	_, _ = b.WriteString("WITH ")
+	for i, cte := range ctes {
+		if i > 0 {
+			_, _ = b.WriteString(",\n")
+		}
+		_, _ = fmt.Fprintf(&b, "%q AS (%s)", cte.As, cte.Query)
+	}
+	_, _ = b.WriteRune('\n')
+	_, _ = b.WriteString(query)
+	return b.String()
+}
+
+// Scan is a convenience wrapper around [CollectOne] that uses [pgx.RowTo] as
+// fn.
+func Scan[T any](
+	ctx context.Context, og *operationGenerator, tx pgx.Tx, query string, args ...any,
+) (result T, err error) {
+	return CollectOne[T](ctx, og, tx, pgx.RowTo[T], query, args...)
+}
+
+// CollectOne is a convenience wrapper around [pgx.CollectOneRow] that logs the
+// query and results via [LogQueryResults]. Usage:
+//
+//	CollectOne(ctx, og, tx, pgx.RowTo[string], `SELECT $1`, "bar")
+func CollectOne[T any](
+	ctx context.Context,
+	og *operationGenerator,
+	tx pgx.Tx,
+	fn pgx.RowToFunc[T],
+	query string,
+	args ...any,
+) (T, error) {
+	rows, err := tx.Query(ctx, query, args...)
+	if err != nil {
+		var zero T
+		return zero, errors.Wrapf(err, "CollectOne: Query: %q %q", query, args)
```

Comment:
Previously, we would log the type of data we were attempting to read through the query which was nominally more information for debugging which we would lose here. Do you see any value in also attempting to log the type of T in `errors.Wrapf`?

Reply:
IIRC, the wrapped error message will contain the type we're scanning into if there's an issue with the scanning portion of the query. If the issue is just with syntax, I don't think adding the type information is particularly useful.